File: Rust/src/ray.rs
use crate::vec3::{Point3, Vec3};

pub struct Ray {
    origin: Point3,
    direction: Vec3,
}

impl Ray {
    pub fn new(origin: Point3, direction: Vec3) -> Self {
        Ray { origin, direction }
    }

    pub fn new_zero() -> Self {
        Ray {
            origin: Point3::new_zero(),
            direction: Vec3::new_zero(),
        }
    }

    pub fn get_origin(&self) -> &Point3 {
        &self.origin
    }

    pub fn get_direction(&self) -> &Vec3 {
        &self.direction
    }

    pub fn at(&self, t: f64) -> Point3 {
        &self.origin + &(&self.direction * t)
    }
}

File: Zig/src/Ray.zig
const Vec3 = @import("Vec3.zig");
pub const Point3 = Vec3;
const expect = @import("std").testing.expect;

orig: Point3,
dir: Vec3,
const zero: @This() = .init(.zero, .zero);

pub fn init(orig: Point3, dir: Vec3) @This() {
    return @This(){
        .orig = orig,
        .dir = dir,
    };
}

pub fn at(self: @This(), t: f64) Point3 {
    return self.orig.add(self.dir.mult(t));
}

test "at" {
    const r = @This().init(.zero, .init(1.0, 2.0, 0.0));
    try expect(r.at(0.4).eql(.init(0.4, 0.8, 0.0)));

    try expect(zero.at(1.0).eql(.zero));

    const r2 = @This().init(.init(1.0, 0.8, 0.0), .init(1.0, 2.0, 0.0));
    try expect(r2.at(1.0).eql(.init(2.0, 2.8, 0.0)));

    // try expect(v1.mult(0.5).eql(@This().init(0.5, 0.0, 1.0)));
}
